<a href="https://colab.research.google.com/github/Soichiro-Gardinner/Model-Validation-Practice/blob/main/%22Imputing_Values_to_prevent_DTLE%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **<font color='#95f1c'>Model Validation Practice</font>**

- What happens before the [<font color='#13f1c'>**Tain Test Split**</font>] ?
- What decitions we make, and why??
- Things to keep in mind.

In [ ]:
# Importing Libraries
import pandas as pd

In [ ]:
# Show all the columns
pd.set_option("display.max_columns", None)

In [ ]:
# Read the file:
df = pd.read_csv("/content/ames-housing-dojo-for-ml.csv")

# Display the Data
df.head(11)

PID MS Zoning  Lot Frontage  Lot Area Street Alley Utilities  \
0   907227090        RL          60.0      7200   Pave   NaN    AllPub   
1   527108010        RL         134.0     19378   Pave   NaN    AllPub   
2   534275170        RL           NaN     12772   Pave   NaN    AllPub   
3   528104050        RL         114.0     14803   Pave   NaN    AllPub   
4   533206070        FV          32.0      3784   Pave  Pave    AllPub   
5   908102320        RL         134.0     17755   Pave   NaN    AllPub   
6   528174080        RL          43.0      7052   Pave   NaN    AllPub   
7   532377130        RL          80.0     10000   Pave   NaN    AllPub   
8   528102030        RL          96.0     12444   Pave   NaN    AllPub   
9   534127170        RL           NaN      9156   Pave   NaN    AllPub   
10  528292080        RL           NaN     12104   Pave   NaN    AllPub   

   Neighborhood Bldg Type House Style  Overall Qual  Overall Cond  Year Built  \
0       CollgCr      1Fam      1Story             5             8        1972   
1       Gilbert      1Fam      2Story             7             5        2005   
2         NAmes      1Fam      1Story             6             8        1960   
3       NridgHt      1Fam      1Story            10             5        2007   
4       Somerst    TwnhsE      1Story             8             5        2006   
5       Edwards      1Fam      1Story             5             4        1959   
6       NridgHt    TwnhsE      1Story             7             5        2005   
7        Sawyer      1Fam      1Story             5             6        1961   
8       NridgHt      1Fam      1Story             8             5        2008   
9        NWAmes      1Fam      1Story             6             7        1968   
10      Gilbert      1Fam      2Story             7             5        2006   

    Year Remodeled Exter Qual Exter Cond  Bsmt Unf Sqft  Total Bsmnt Sqft  \
0             1972         TA         TA          427.0             864.0   
1             2006         Gd         TA         1335.0            1392.0   
2             1998         TA         Gd          460.0             958.0   
3             2008         Ex         TA          442.0            2078.0   
4             2007         Gd         TA         1451.0            1511.0   
5             1959         TA         TA         1290.0            1466.0   
6             2005         Gd         TA          705.0            1364.0   
7             1983         TA         TA          270.0             864.0   
8             2008         Ex         TA          596.0            1932.0   
9             1968         TA         TA         1489.0            1489.0   
10            2006         Gd         TA         1035.0            1035.0   

   Central Air  Living Area Sqft  Bedroom  Kitchen  Total Rooms Garage Type  \
0            Y             864.0        3        1            5      Detchd   
1            Y            2462.0        4        1            9      Attchd   
2            Y             958.0        2        1            5      Attchd   
3            Y            2084.0        2        1            7      Attchd   
4            Y            1565.0        2        1            5      Attchd   
5            Y            1466.0        3        1            6      Attchd   
6            Y            1364.0        2        1            6      Attchd   
7            Y            1144.0        3        1            6      Attchd   
8            Y            1932.0        2        1            7      Attchd   
9            Y            1489.0        3        1            7      Attchd   
10           Y            2322.0        4        1            8      Attchd   

    Garage Yr Blt  Garage Cars  Garage Area Garage Qual Garage Cond  \
0          1977.0          1.0        297.0          TA          TA   
1          2006.0          2.0        576.0          TA          TA   
2          1960.0          1.0        301.0          TA      

In [ ]:
df.columns

Index(['PID', 'MS Zoning', 'Lot Frontage', 'Lot Area', 'Street', 'Alley',
       'Utilities', 'Neighborhood', 'Bldg Type', 'House Style', 'Overall Qual',
       'Overall Cond', 'Year Built', 'Year Remodeled', 'Exter Qual',
       'Exter Cond', 'Bsmt Unf Sqft', 'Total Bsmnt Sqft', 'Central Air',
       'Living Area Sqft', 'Bedroom', 'Kitchen', 'Total Rooms', 'Garage Type',
       'Garage Yr Blt', 'Garage Cars', 'Garage Area', 'Garage Qual',
       'Garage Cond', 'Paved Drive', 'Fence', 'SalePrice', 'Month', 'Year',
       'Total Full Baths', 'Total Half Baths'],
      dtype='object')

In [ ]:
# How many NaNs on each row?
print("-- --There are-- --")
df.isnull().sum()

-- --There are-- --


PID                    0
MS Zoning              0
Lot Frontage         490
Lot Area               0
Street                 0
Alley               2732
Utilities              0
Neighborhood           0
Bldg Type              0
House Style            0
Overall Qual           0
Overall Cond           0
Year Built             0
Year Remodeled         0
Exter Qual             0
Exter Cond             0
Bsmt Unf Sqft          1
Total Bsmnt Sqft       1
Central Air            0
Living Area Sqft       0
Bedroom                0
Kitchen                0
Total Rooms            0
Garage Type          157
Garage Yr Blt        160
Garage Cars            1
Garage Area            1
Garage Qual          159
Garage Cond          159
Paved Drive            0
Fence               2358
SalePrice              0
Month                  0
Year                   0
Total Full Baths       2
Total Half Baths       5
dtype: int64

In [ ]:
# How many Rows and Columns we have?
print(f"Rows = {df.shape[0]}, Columns = {df.shape[1]}")

Rows = 2930, Columns = 36


- Look how some columns have [2,000 NaNs] on them while the Dataset shape is [2,930 Rows] and [36 Columns].

- **What Would you do in this situation?**

____________________________________

# **<font color='#95f1c'>1. Drop redundant columns**

- <font color='#1ff1c'>I will drop the columns with more than 2,000 NaNs because they do not have relevant data on them.</font>
- Even if I try to fill them using the **[.mean() or .mode()]** it will be using more NaNs to do the math than real numbers or information.

In [ ]:
# drop irrelevant columns
df = df.drop(columns = "Fence")
df = df.drop(columns = "Alley")

# Display how it looks now:
df.head()

PID MS Zoning  Lot Frontage  Lot Area Street Utilities Neighborhood  \
0  907227090        RL          60.0      7200   Pave    AllPub      CollgCr   
1  527108010        RL         134.0     19378   Pave    AllPub      Gilbert   
2  534275170        RL           NaN     12772   Pave    AllPub        NAmes   
3  528104050        RL         114.0     14803   Pave    AllPub      NridgHt   
4  533206070        FV          32.0      3784   Pave    AllPub      Somerst   

  Bldg Type House Style  Overall Qual  Overall Cond  Year Built  \
0      1Fam      1Story             5             8        1972   
1      1Fam      2Story             7             5        2005   
2      1Fam      1Story             6             8        1960   
3      1Fam      1Story            10             5        2007   
4    TwnhsE      1Story             8             5        2006   

   Year Remodeled Exter Qual Exter Cond  Bsmt Unf Sqft  Total Bsmnt Sqft  \
0            1972         TA         TA          427.0             864.0   
1            2006         Gd         TA         1335.0            1392.0   
2            1998         TA         Gd          460.0             958.0   
3            2008         Ex         TA          442.0            2078.0   
4            2007         Gd         TA         1451.0            1511.0   

  Central Air  Living Area Sqft  Bedroom  Kitchen  Total Rooms Garage Type  \
0           Y             864.0        3        1            5      Detchd   
1           Y            2462.0        4        1            9      Attchd   
2           Y             958.0        2        1            5      Attchd   
3           Y            2084.0        2        1            7      Attchd   
4           Y            1565.0        2        1            5      Attchd   

   Garage Yr Blt  Garage Cars  Garage Area Garage Qual Garage Cond  \
0         1977.0          1.0        297.0          TA          TA   
1         2006.0          2.0        576.0          TA          TA   
2         1960.0          1.0        301.0          TA          TA   
3         2007.0          3.0       1220.0          TA          TA   
4         2006.0          2.0        476.0          TA          TA   

  Paved Drive  SalePrice  Month  Year  Total Full Baths  Total Half Baths  
0           Y   119900.0      3  2006               1.0               0.0  
1           Y   320000.0      3  2006               3.0               1.0  
2           Y   151500.0      4  2007               1.0               0.0  
3           Y   385000.0      6  2008               3.0               0.0  
4           Y   193800.0      2  2007               3.0               0.0

In [ ]:
print(f"Columns = {df.shape[1]}")

Columns = 34


- **What's next?**

- **Should I drop the duplicates or should I fill the NaNs?**

____________________

# **<font color='#95f1c'>2. Drop Duplicates**

- We will Drop Duplicates Because we still don't understand our data %100 percent.
- This will also allow us to clean some NaNs that were created by accident.

In [ ]:
# Drop Duplicates
df = df.drop_duplicates()

In [ ]:
print(f"Rows = {df.shape[0]}")  # The Ouput showed that we don't have any Duplicates on this Dataset

Rows = 2930


_______________

# **<font color='#95f1c'>3. Analyze Data and Clean NaNs**


In [ ]:
# Check the Data types:
df.dtypes

PID                   int64
MS Zoning            object
Lot Frontage        float64
Lot Area              int64
Street               object
Utilities            object
Neighborhood         object
Bldg Type            object
House Style          object
Overall Qual          int64
Overall Cond          int64
Year Built            int64
Year Remodeled        int64
Exter Qual           object
Exter Cond           object
Bsmt Unf Sqft       float64
Total Bsmnt Sqft    float64
Central Air          object
Living Area Sqft    float64
Bedroom               int64
Kitchen               int64
Total Rooms           int64
Garage Type          object
Garage Yr Blt       float64
Garage Cars         float64
Garage Area         float64
Garage Qual          object
Garage Cond          object
Paved Drive          object
SalePrice           float64
Month                 int64
Year                  int64
Total Full Baths    float64
Total Half Baths    float64
dtype: object

In [ ]:
# Now Look at the Data Again and Make sure you understand what kind of data
# you have and what they represent.

# - Are They in the correct Dtype?

In [ ]:
df.head(11)

PID MS Zoning  Lot Frontage  Lot Area Street Utilities Neighborhood  \
0   907227090        RL          60.0      7200   Pave    AllPub      CollgCr   
1   527108010        RL         134.0     19378   Pave    AllPub      Gilbert   
2   534275170        RL           NaN     12772   Pave    AllPub        NAmes   
3   528104050        RL         114.0     14803   Pave    AllPub      NridgHt   
4   533206070        FV          32.0      3784   Pave    AllPub      Somerst   
5   908102320        RL         134.0     17755   Pave    AllPub      Edwards   
6   528174080        RL          43.0      7052   Pave    AllPub      NridgHt   
7   532377130        RL          80.0     10000   Pave    AllPub       Sawyer   
8   528102030        RL          96.0     12444   Pave    AllPub      NridgHt   
9   534127170        RL           NaN      9156   Pave    AllPub       NWAmes   
10  528292080        RL           NaN     12104   Pave    AllPub      Gilbert   

   Bldg Type House Style  Overall Qual  Overall Cond  Year Built  \
0       1Fam      1Story             5             8        1972   
1       1Fam      2Story             7             5        2005   
2       1Fam      1Story             6             8        1960   
3       1Fam      1Story            10             5        2007   
4     TwnhsE      1Story             8             5        2006   
5       1Fam      1Story             5             4        1959   
6     TwnhsE      1Story             7             5        2005   
7       1Fam      1Story             5             6        1961   
8       1Fam      1Story             8             5        2008   
9       1Fam      1Story             6             7        1968   
10      1Fam      2Story             7             5        2006   

    Year Remodeled Exter Qual Exter Cond  Bsmt Unf Sqft  Total Bsmnt Sqft  \
0             1972         TA         TA          427.0             864.0   
1             2006         Gd         TA         1335.0            1392.0   
2             1998         TA         Gd          460.0             958.0   
3             2008         Ex         TA          442.0            2078.0   
4             2007         Gd         TA         1451.0            1511.0   
5             1959         TA         TA         1290.0            1466.0   
6             2005         Gd         TA          705.0            1364.0   
7             1983         TA         TA          270.0             864.0   
8             2008         Ex         TA          596.0            1932.0   
9             1968         TA         TA         1489.0            1489.0   
10            2006         Gd         TA         1035.0            1035.0   

   Central Air  Living Area Sqft  Bedroom  Kitchen  Total Rooms Garage Type  \
0            Y             864.0        3        1            5      Detchd   
1            Y            2462.0        4        1            9      Attchd   
2            Y             958.0        2        1            5      Attchd   
3            Y            2084.0        2        1            7      Attchd   
4            Y            1565.0        2        1            5      Attchd   
5            Y            1466.0        3        1            6      Attchd   
6            Y            1364.0        2        1            6      Attchd   
7            Y            1144.0        3        1            6      Attchd   
8            Y            1932.0        2        1            7      Attchd   
9            Y            1489.0        3        1            7      Attchd   
10           Y            2322.0        4        1            8      Attchd   

    Garage Yr Blt  Garage Cars  Garage Area Garage Qual Garage Cond  \
0          1977.0          1.0        297.0          TA          TA   
1          2006.0          2.0        576.0          TA          TA   
2          1960.0          1.0        301.0          TA          TA   
3          2007.0          3.0       1220.0          TA       

- As you can see there is some strange data I don't really understand.

#### **For Example**:
     The "MS zoning" column, and the "Exter Qual".
     
<font color='#13f1c'>I Don't even understand the Names to be honest, I'm new to this.

## **In such cases you would like to ask for the Data Dictionary**
- ask ChatGPT or do some research online.

<font color='#13f1c'>but in this case we don't have it so after making sure they are in the correct datatype we will finally fill the NaNs

___________________

# <font color='Red'> **Attention**</font>:
- ## **<font color='Black'> We only fill the NaNs Separetly after the Split to prevent data leakage**


___________________

# **Data Leakage**
- Data Leakage basically means that data from our "training set" affected our "test set" or that we used data from our "training set" to change data from our "testing set".

# **<font color='#95f1c'>4. Train_Test_Split**

In [ ]:
# import the train_test_split function:

from sklearn.model_selection import train_test_split as TTS # Some people would not rename it but I did it to make it short.

In [ ]:
# Separate the Targe Value column (What we will predict).
y = df["SalePrice"]


# Now Separate the Features (The data our computer will learn in order to predict)
X = df.drop(columns = "SalePrice")

- we have saved our target in the variable "y".
- we have saved our features excluding our target in the var "X"

In [ ]:
# Now we do the split:
X_train, X_test, y_train, y_test = TTS(X,y, random_state = 42) # random_state = 42 its kind of the most used number to split
                                                               ## but there is also a formula to find the best random_state.

- Note how we will get 4 variables from the 2 "X" & "y" we created before.

- The varibles were used by the "TTS function" to identify the target/features and separate them into 'training' and 'testing set'.

# <font color='#95f1c'>**5. Imputing values**
- **What happens Before?**
- **What is the process?**

In [ ]:
# Select Numeric Columns
num_cols = X_train.select_dtypes("number").columns
num_cols

Index(['PID', 'Lot Frontage', 'Lot Area', 'Overall Qual', 'Overall Cond',
       'Year Built', 'Year Remodeled', 'Bsmt Unf Sqft', 'Total Bsmnt Sqft',
       'Living Area Sqft', 'Bedroom', 'Kitchen', 'Total Rooms',
       'Garage Yr Blt', 'Garage Cars', 'Garage Area', 'Month', 'Year',
       'Total Full Baths', 'Total Half Baths'],
      dtype='object')

- **.select_dtypes("number").columns** will select the columns that hold numeric values.

### **Start the Imputer**

In [ ]:
from sklearn.impute import SimpleImputer

# start the inputer with the "median" strategy:
impute_median = SimpleImputer(strategy="median")

In [ ]:
# Fit the object on numeric training data with .fit():
impute_median.fit(X_train[num_cols])

SimpleImputer(strategy='median')

In [ ]:
# check how many Null/NaNs are in the X_train numeric columns:
null_count = X_train[num_cols].isna().sum()
null_count

PID                   0
Lot Frontage        366
Lot Area              0
Overall Qual          0
Overall Cond          0
Year Built            0
Year Remodeled        0
Bsmt Unf Sqft         0
Total Bsmnt Sqft      0
Living Area Sqft      0
Bedroom               0
Kitchen               0
Total Rooms           0
Garage Yr Blt       125
Garage Cars           1
Garage Area           1
Month                 0
Year                  0
Total Full Baths      1
Total Half Baths      3
dtype: int64

if you want to see only the ones that have nulls do this:

In [ ]:
null_count[null_count > 0] # filtering columns that have more than 0 nulls/NaNs

Lot Frontage        366
Garage Yr Blt       125
Garage Cars           1
Garage Area           1
Total Full Baths      1
Total Half Baths      3
dtype: int64

Notice that fitting the inputer didn't work, It is because we need to use the **.transform()** to really apply the imputer strategy. So everytime you think about imputing thing about **"transforming by imputing"**

In [ ]:
# Inpute Correctly:
X_train_numeric_Imputed = impute_median.transform(X_train[num_cols])
X_train_numeric_Imputed

array([[9.0547552e+08, 6.8000000e+01, 1.1515000e+04, ..., 2.0070000e+03,
        1.0000000e+00, 0.0000000e+00],
       [9.0925401e+08, 5.3000000e+01, 7.1550000e+03, ..., 2.0090000e+03,
        1.0000000e+00, 0.0000000e+00],
       [5.3145009e+08, 6.8000000e+01, 7.1530000e+03, ..., 2.0080000e+03,
        3.0000000e+00, 0.0000000e+00],
       ...,
       [5.2821809e+08, 6.3000000e+01, 1.0928000e+04, ..., 2.0090000e+03,
        2.0000000e+00, 1.0000000e+00],
       [5.3535313e+08, 6.8000000e+01, 1.5783000e+04, ..., 2.0070000e+03,
        1.0000000e+00, 0.0000000e+00],
       [9.0913117e+08, 6.8000000e+01, 1.2155000e+04, ..., 2.0080000e+03,
        2.0000000e+00, 0.0000000e+00]])

**Is that what we wanted?**
- <font color='Red'> No, it is an array, and we need a Dataframe.

In [ ]:
# Change settings to ouput a dataframe
from sklearn import set_config
set_config(transform_output='pandas')

In [ ]:
# Transform the training data (fill in nulls with median)
X_train_num_imputed = impute_median.transform(X_train[num_cols])
X_train_num_imputed.head()

PID  Lot Frontage  Lot Area  Overall Qual  Overall Cond  \
844   905475520.0          68.0   11515.0           4.0           5.0   
2730  909254010.0          53.0    7155.0           7.0           8.0   
2793  531450090.0          68.0    7153.0           6.0           5.0   
1187  903400040.0          60.0   13515.0           6.0           6.0   
2770  527107130.0          60.0    8400.0           7.0           5.0   

      Year Built  Year Remodeled  Bsmt Unf Sqft  Total Bsmnt Sqft  \
844       1958.0          1994.0            0.0               0.0   
2730      1918.0          1990.0          600.0             600.0   
2793      1991.0          1991.0           78.0            1278.0   
1187      1919.0          1950.0          764.0             764.0   
2770      1996.0          1997.0          100.0             384.0   

      Living Area Sqft  Bedroom  Kitchen  Total Rooms  Garage Yr Blt  \
844              943.0      3.0      1.0          5.0         1958.0   
2730            1228.0      3.0      1.0          6.0         1918.0   
2793            1294.0      3.0      1.0          6.0         1991.0   
1187            1824.0      3.0      1.0          8.0         1940.0   
2770            1628.0      3.0      1.0          7.0         1996.0   

      Garage Cars  Garage Area  Month    Year  Total Full Baths  \
844           1.0        308.0    8.0  2007.0               1.0   
2730          1.0        215.0    5.0  2009.0               1.0   
2793          2.0        496.0    6.0  2008.0               3.0   
1187          2.0        520.0    7.0  2007.0               1.0   
2770          2.0        390.0    6.0  2009.0               2.0   

      Total Half Baths  
844                0.0  
2730               0.0  
2793               0.0  
1187               0.0  
2770               1.0

In [ ]:
X_train_num_imputed.isna().sum()

PID                 0
Lot Frontage        0
Lot Area            0
Overall Qual        0
Overall Cond        0
Year Built          0
Year Remodeled      0
Bsmt Unf Sqft       0
Total Bsmnt Sqft    0
Living Area Sqft    0
Bedroom             0
Kitchen             0
Total Rooms         0
Garage Yr Blt       0
Garage Cars         0
Garage Area         0
Month               0
Year                0
Total Full Baths    0
Total Half Baths    0
dtype: int64

## Imputing Categorical and Ordinal columns/features


In [ ]:
# Define the "Categorical features" variable
cat_fish = X_train.select_dtypes("object").columns
cat_fish

Index(['MS Zoning', 'Street', 'Utilities', 'Neighborhood', 'Bldg Type',
       'House Style', 'Exter Qual', 'Exter Cond', 'Central Air', 'Garage Type',
       'Garage Qual', 'Garage Cond', 'Paved Drive'],
      dtype='object')

In [ ]:
# Start the imputer:
cat_imputer = SimpleImputer(strategy = "constant", fill_value = "NA")

In [ ]:
# Fit the imputer on Training Data ONLY!!
cat_imputer.fit(X_train[cat_fish])

SimpleImputer(fill_value='NA', strategy='constant')

- **Tools you could use:**

In [ ]:
#Check what stragy U used:
print("The Stragy U used is:",cat_imputer.strategy.upper())    # .upper() just make any string output appear in UpperCase letters

The Stragy U used is: CONSTANT


In [ ]:
#Check What values you are going to impute:
print("The Stragy U used is:",cat_imputer.fill_value.lower())

The Stragy U used is: na


In [ ]:
#Check what columns are you going to impute:
print("Columns you \".fit()\" on the imputer:",cat_imputer.feature_names_in_)

Columns you ".fit()" on the imputer: ['MS Zoning' 'Street' 'Utilities' 'Neighborhood' 'Bldg Type' 'House Style'
 'Exter Qual' 'Exter Cond' 'Central Air' 'Garage Type' 'Garage Qual'
 'Garage Cond' 'Paved Drive']


In [ ]:
# NaNs in Our Columns now?
X_train[cat_fish].isna().sum().sum()

370

In [ ]:
# Transform the NaNs/Nulls and save it in a new variable:
X_train_imputed = cat_imputer.transform(X_train[cat_fish])

In [ ]:
# Look for Nulls on the new X_train variable you created
X_train_imputed.isna().sum().sum()

0

### Now there are 0 Nulls/NaNs on the "numeric" and "categorical" data.

# <font color='#95f1c'>**6. Ordinal and OneHotEncoder**
- **What do they do?**
- **Why do we use them?**

## **Ordinal**

In [ ]:
# define a list of columns to encode as ordinal
ordinal_cols = ['Exter Qual', 'Garage Qual', 'Exter Cond', 'Garage Cond']
ordinal_cols

['Exter Qual', 'Garage Qual', 'Exter Cond', 'Garage Cond']

# **<font color='#95f1c'>8. Scaling and Standarization**

- **Scaling**: In the context of data science and machine learning, the term "scale" refers to the range of the data, which is the difference between the maximum and minimum values within a particular feature (column).


- **Standariztion**:is a preprocessing technique used to transform features (variables) to have a mean of zero and a standard deviation of one. This transformation results in the feature having properties of a standard normal distribution.